hazelcast/src/main/java/com/hazelcast/client/impl/operations/ClientDisconnectionOperation.java

```diff
@@ -46,28 +44,10 @@ public ClientDisconnectionOperation(String clientUuid, String memberUuid) {
     @Override
     public void run() throws Exception {
         ClientEngineImpl engine = getService();
-        final ClientEndpointManagerImpl endpointManager = (ClientEndpointManagerImpl) engine.getEndpointManager();
-        if (!engine.removeOwnershipMapping(clientUuid, memberUuid)) {
-            return;
-        }
-
-        Set<ClientEndpoint> endpoints = endpointManager.getEndpoints(clientUuid);
-        // This part cleans up listener and transactions
-        for (ClientEndpoint endpoint : endpoints) {
-            endpoint.getConnection().close("ClientDisconnectionOperation: Client disconnected from cluster", null);
-        }
-
-        // This part cleans up locks conditions semaphore etc..
         NodeEngineImpl nodeEngine = (NodeEngineImpl) getNodeEngine();
-        Collection<ClientAwareService> services = nodeEngine.getServices(ClientAwareService.class);
-        for (ClientAwareService service : services) {
-            service.clientDisconnected(clientUuid);
-        }
-    }
-
-    @Override
-    public boolean returnsResponse() {
-        return false;
+        Callable<Boolean> task = new ClientDisconnectedTask(engine, nodeEngine, clientUuid, memberUuid);
+        Future<Boolean> future = engine.getClientManagementExecutor().submit(task);
+        future.get();
```

Comment:
I think this approach can lead to deadlocks. Here, we are on a generic operation thread, we offload a task to an executor thread and wait for its result. However, in `com.hazelcast.transaction.impl.TransactionManagerServiceImpl#clientDisconnected()`, we make new invocations to rollback transactions, which will run on generic operation threads as well. When we have a sufficient number of client disconnections, we can occupy all op threads and fall into a deadlock...